apps/src/javalab/Javalab.js

```diff
@@ -92,13 +101,23 @@ Javalab.prototype.init = function(config) {
   });
 
   registerReducers({javalab});
-
-  ReactDOM.render(
-    <Provider store={getStore()}>
-      <JavalabView onMount={onMount} onContinue={onContinue} />
-    </Provider>,
-    document.getElementById(config.containerId)
-  );
+  const onSave = this.onSave.bind(this);
+  const renameFile = this.renameProjectFile.bind(this);
+
+  // only render Java Lab once files have been loaded.
+  this.loadFiles(() => {
```

Comment:
Rather than passing this down multiple levels as a callback, would it make more sense to make it a member function of Javalab and call it directly? Alternatively, consider renaming the `callback` parameter in `loadFiles`, `populateFiles`, and `requestFileEntryAndWrite` to something more descriptive.

Reply:
combining this feedback with your module idea below--refactoring the file management into a module and importing the functions into the required components (and using `success` and `failure` instead of `callback`). I think keeping all the file management in one place is a good idea because it will get more complicated with multi-file.

Reply:
one downside of moving the load into the view is we are losing Studio App's default behavior of showing a "loading is taking a long time" message with a reload link if the page load takes a long time--we can discuss if we want to add that back or not, not super important right now.

Reply:
That message would be nice to have, but I don't think it should block this PR. Why do we lose the message?

Reply:
Refactor looks great :)

Reply:
the message pops up when the `init` function takes a certain amount of time--since we are loading after init completes now we lose the message.